Amazon Spearman and Spear Weapon

TagnumElite opened the pull request:
I have created an Amazon Spearman as well as a Spear item that acts just like a trident, just with out riptide.

# Changes proposed in this pull request:
- Add Amazon Spearman Entity
- Add Spear Weapon
- Allow ranged mobs to also shoot spears and tridents

Review please
The Spear:
![spear pictures](https://user-images.githubusercontent.com/6497721/138592847-7a1284b9-6b69-48d3-84b1-d0bc878d92f8.png)
The Amazon Spearman (is a link to a video)
[![amazon pictures](https://user-images.githubusercontent.com/6497721/138761061-84ddc850-d6ce-4e1e-9864-0acc6a45f654.png)](https://youtu.be/nXfafa6Er64)

TagnumElite committed:
Create an Amazon Spearman

Duplicate of amazon archer model as I am not an good enough artist.

TagnumElite committed:
Add spear item

Move FireArrow render in projectiles subpackage
Add missing mob amazon mob attributes

TagnumElite committed:
Make spearman use spears and also add trident and spear support for mobs

Raycoms commented:
Can we have some images?

Raycoms commented on src/api/java/com/minecolonies/api/entity/SpearEntity.java:
```diff
@@ -0,0 +1,246 @@
+    @Override
+    public void tick()
+    {
+        if (this.inGroundTime > 4)
```
Does all of this here have to happen every tick?

Raycoms commented on src/api/java/com/minecolonies/api/entity/SpearEntity.java:
```diff
@@ -0,0 +1,246 @@
+
+                this.remove();
+            }
+            else if (i > 0)
```
What is i?

Raycoms commented on src/api/java/com/minecolonies/api/entity/SpearEntity.java:
```diff
@@ -0,0 +1,246 @@
+            }
+        }
+
+        this.setDeltaMovement(this.getDeltaMovement().multiply(-0.01D, -0.1D, -0.01D));
```
I think we don't need the thundering bonus here. Doesn't fit the amazons too well.

Raycoms commented on src/api/java/com/minecolonies/api/entity/mobs/RaiderMobUtils.java:
```diff
@@ -163,6 +163,9 @@ else if (mob instanceof IArcherMobEntity)
         }
+        else if (mob instanceof ISpearmanMobEntity) {
+            mob.setItemSlot(EquipmentSlotType.MAINHAND, new ItemStack(ModItems.spear));
+        }
```
formatting

Raycoms commented on src/main/java/com/minecolonies/coremod/client/render/projectile/RendererSpear.java:
```diff
@@ -0,0 +1,57 @@
+    private final ResourceLocation texture;
+    private final ModelSpear model = new ModelSpear();
+
+    public RendererSpear(EntityRendererManager entityRendererManager) {
```
formatting

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/colonyEvents/raidEvents/amazonevent/AmazonRaidEvent.java:
```diff
@@ -99,14 +102,6 @@ public void onUpdate()
         }
     }
 
-    @Override
```
I don't think you should remove those two methods

TagnumElite replied:
The were overridden because there was no normal raider type for amazons. I have set the normal mob type to spearmen, therefore these overrides are no longer needed.

Raycoms replied:
ah okay,

TagnumElite committed:
Fix the model not spawning when thrown.

To be honest, don't know why this fixes it, but it does.

marchermans commented on src/main/java/com/minecolonies/coremod/entity/ai/registry/MobAIRegistry.java:
```diff
@@ -48,11 +49,11 @@ private static void setupMobAiTasks(final IMobAIRegistry registry)
           .registerNewAiTaskForMobs(PRIORITY_SIX, mob -> new LookAtGoal(mob, EntityCitizen.class, MAX_WATCH_DISTANCE))
           .registerNewStateAI(mob -> new RaiderMeleeAI<>(mob, mob.getAI()), mob -> !(mob instanceof IArcherMobEntity))
-          .registerNewStateAI(mob -> new RaiderRangedAI<>(mob, mob.getAI()), mob -> mob instanceof IArcherMobEntity)
+          .registerNewStateAI(mob -> new RaiderRangedAI<>(mob, mob.getAI()), mob -> mob instanceof IArcherMobEntity || mob instanceof ISpearmanMobEntity)
```
Should be collapsed into a single super interface: `IRangedAttackEntity` instead of having two different interfaces for this.

TagnumElite replied:
Will do

marchermans commented on src/main/java/com/minecolonies/coremod/items/ItemSpear.java:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    @OnlyIn(Dist.CLIENT)
+    public static class SpearTileRenderer extends ItemStackTileEntityRenderer
```
Please do not mix client code in inner classes.
This is a disaster waiting to happen on servers.

Extract this class and move it to the correct client package.

TagnumElite replied:
Yeah, will do.

marchermans reviewed (changes requested):
Generally a good start.
I agree with Ray that some code likely does not need to run every tick on the spear entity.
And some clean up needs to be done before this is ready.

But all in all a good start

TagnumElite commented:
Yeah, thats why I left this as a draft pr. I am using the trident as a reference, but I was thinking. The spear doesn't need the loyal enchant, just has to be generic right?

TagnumElite committed:
Use an IRangedMovEntity interface

Makes Archer and Spearmen extend from ranged.
I didn't change the check for melee checking for ArcherMobEntity because I want the spear to be used as both melee and ranged.

TagnumElite committed:
Move SpearTileRenderer to its own class

Also, cleanup ItemSpear class

TagnumElite committed:
Simplify SpearEntity

Removed some overrides that weren't needed.
Removed ID_LOYALTY because I don't think the spear should have it.

Raycoms commented:
> Can we have some images?

Also one of the spearman model?

TagnumElite commented:
I duplicated the normal amazon model. Will be making a new model today.

Raycoms commented:
> I duplicated the normal amazon model. Will be making a new model today.

For me using the current archer model is totally fine, I'd just slightly adjust the skin a bit.

TagnumElite committed:
Update Spearman Model and Texture

TagnumElite committed:
Fix model rendering

I just trusted BlockBench, nope.

TagnumElite committed:
Create an ICustomAttackSound interface

This interface is used by RaiderRangedAI to use for attacking with a different sound.

TagnumElite committed:
Fix some formatting

TagnumElite commented:
I believe that is everything, only thing I dislike right now is the spear item texture. I'm not an artist.

Raycoms commented:
@TagnumElite is this ready for review? You need to remove the not ready tag then?

TagnumElite commented:
I am unable to edit ~tags~ labels, which is why I sent the message.

someaddons commented on src/api/java/com/minecolonies/api/entity/SpearEntity.java:
```diff
@@ -0,0 +1,154 @@
+import net.minecraftforge.fml.network.NetworkHooks;
+import org.jetbrains.annotations.NotNull;
+
+public class SpearEntity extends AbstractArrowEntity implements ICustomAttackSound
```
missing javadoc on class and vars

someaddons replied:
The projectile should get a limited lifetime similar to our existing CustomArrowEntity, to make sure it'll despawn, even if it never hits the ground

someaddons commented on src/api/java/com/minecolonies/api/entity/mobs/amazons/IAmazonSpearman.java:
```diff
@@ -0,0 +1,8 @@
+
+import com.minecolonies.api.entity.mobs.ISpearmanMobEntity;
+
+public interface IAmazonSpearman extends IAmazonEntity, ISpearmanMobEntity
```
javadoc missing

someaddons commented on src/main/java/com/minecolonies/coremod/client/model/ModelSpear.java:
```diff
@@ -0,0 +1,61 @@
+import org.jetbrains.annotations.NotNull;
+
+@OnlyIn(Dist.CLIENT)
+public class ModelSpear extends Model
```
javadoc missing in this class

someaddons reviewed (commented):
Looks good, just some doc/comments left

TagnumElite committed:
Add missing documentation

TagnumElite committed:
Implement timed lifetime for the spear entity

This sets it to be removed after 60 seconds.
